### PR TITLE
Client/Jetpack-Connect: Migrate `authorize`'s tests to RTL

### PR DESCRIPTION
**Pull request by fullofcaffeine**
Tracking issue: https://github.com/Automattic/wp-calypso/issues/63409.

#### Proposed Changes

* Migrate `cient/jetpack-connect/authorize`'s component's tests to RTL, and remove Enzime's references from it.

#### Testing Instructions

* Tests should pass.

#### Pre-merge Checklist

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to #

### CODE REVIEW

**Comment by fullofcaffeine** [comment]
This is a tricky one. It's still a [class React component](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-connect/authorize.js) and relies heavily on accessing and testing public members of the class instance directly ([example](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-connect/test/authorize.js#L85-L96)). I think it's okay as this kind of hybrid component that's mainly non-visual, with the visual part for the end user being only a small part of it, as far as I understand, so far.

I think it's worth discussing how to refactor the component and the tests. It'd be overkill and risky to do that as part of this PR, given that I have never worked with this component before, and it's a crucial component for JP. It'd probably make sense to refactor this to a function component as the next step. 

TL;DR: So, given its complexity, in this initial iteration, I'll only migrate tests that do some kind of rendering of the component. I'll leave out other examples that check props and public methods directly, as-is. Hopefully, this should at least clean up the test of enzyme references.

cc @Automattic/team-calypso-frameworks Thoughts?

**Comment by tyxla** [comment]
> This is a tricky one. It's still a [class React component](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-connect/authorize.js) and relies heavily on accessing and testing public members of the class instance directly ([example](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-connect/test/authorize.js#L85-L96)). I think it's okay as this kind of hybrid component that's mainly non-visual, with the visual part for the end user being only a small part of it, as far as I understand, so far.
> 
> I think it's worth discussing how to refactor the component and the tests. It'd be overkill and risky to do that as part of this PR, given that I have never worked with this component before, and it's a crucial component for JP. It'd probably make sense to refactor this to a function component as the next step.
> 
> TL;DR: So, given its complexity, in this initial iteration, I'll only migrate tests that do some kind of rendering of the component. I'll leave out other examples that check props and public methods directly, as-is. Hopefully, this should at least clean up the test of enzyme references.
> 
> cc @Automattic/team-calypso-frameworks Thoughts?

I agree 👍 Having worked on the original components in Jetpack Connect before, I'd suggest going above and beyond to keep the current logic of the tests. Much of the logic is complex, many flows are likely unreachable and unused at this point, so forcing the component to get to those flows could be difficult. In favor of finishing the migration, I'd suggest keeping the tests as they are as much as possible, and just migrating away from `enzyme` usage. We could spend some time in the future cleaning up the codebase of unused flows and modernize it to use functional components and hooks, but this should be a separate effort backed up by a solid plan with lots of testing involved. 

That being said, feel free to ping me if you have any concerns or blockers while migrating the tests here.

File: client/jetpack-connect/test/authorize.js [hunk 1]
```diff
@@ -63,6 +69,24 @@ const DEFAULT_PROPS = deepFreeze( {
...
+jest.mock( '../persistence-utils', () => ( {
+	...jest.requireActual( '../persistence-utils' ),
+	isSsoApproved: ( clientId ) => clientId === APPROVE_SSO_CLIENT_ID,
+} ) );
+
+function renderWithRedux( ui ) {
+	return renderWithProvider( ui, {
+		reducers: {
+			ui: uiReducer,
+			documentHead: documentHeadReducer,
```
Comment (tyxla) [inline client/jetpack-connect/test/authorize.js:83]: We could also mock `<DocumentHead />` but this should do it just equally well too 👍

File: client/jetpack-connect/test/authorize.js [hunk 2]
```diff
@@ -76,9 +100,9 @@ jest.mock( '@automattic/calypso-config', () => {
...
 describe( 'JetpackAuthorize', () => {
 	test( 'renders as expected', () => {
-		const wrapper = shallow( <JetpackAuthorize { ...DEFAULT_PROPS } /> );
+		const { container } = renderWithRedux( <JetpackAuthorize { ...DEFAULT_PROPS } /> );
 
-		expect( wrapper ).toMatchSnapshot();
+		expect( container ).toMatchSnapshot();
 	} );
 
 	describe( 'isSso', () => {
```
Comment (tyxla) [inline client/jetpack-connect/test/authorize.js:108]: Not necessary for this PR, as this is pre-existing, but if this can be tested without rendering, it's usually a good candidate for extracting to a separate utility function.

File: client/jetpack-connect/test/authorize.js [hunk 3]
```diff
@@ -155,55 +179,81 @@ describe( 'JetpackAuthorize', () => {
...
-			const component = shallow( renderableComponent );
-			component.setProps( {
-				authQuery: {
-					...DEFAULT_PROPS.authQuery,
-					from: 'woocommerce-setup-wizard',
-				},
-			} );
-			const result = component.instance().shouldAutoAuthorize();
+		test( 'should auto-authorize for the old WOO setup wizard', () => {
+			const authorizeMock = jest.fn();
```
Comment (tyxla) [inline client/jetpack-connect/test/authorize.js]: Optionally we could create this mock on `beforeAll()` and reset it on `afterEach()`, or just create it on `beforeEach()` to avoid boilerplate in every test.

File: client/jetpack-connect/test/authorize.js [hunk 4]
```diff
@@ -304,46 +354,107 @@ describe( 'JetpackAuthorize', () => {
...
-			const component = shallow( renderableComponent );
-			component.setProps( {
-				partnerSlug: 'pressable',
-			} );
-			const target = component.instance().getRedirectionTarget();
+		let originalWindowLocation;
+
+		beforeEach( () => {
+			originalWindowLocation = global.window.location;
+			delete global.window.location;
```
Comment (tyxla) [inline client/jetpack-connect/test/authorize.js:359]: I might be missing something, but since `global.window.location` is an object, `originalWindowLocation ` will lead to it, and if we `delete global.window.location`, it will also erase `originalWindowLocation`, won't it?

Also, why do we need to `delete` it if we overwrite it on the next line?

File: client/jetpack-connect/test/authorize.js [hunk 5]
```diff
@@ -304,46 +354,107 @@ describe( 'JetpackAuthorize', () => {
...
+		test( 'should redirect to pressable if partnerSlug is "pressable"', async () => {
+			renderWithRedux(
+				<JetpackAuthorize
+					{ ...DEFAULT_PROPS }
+					authQuery={ {
+						...DEFAULT_PROPS.authQuery,
+						alreadyAuthorized: true,
+					} }
+					partnerSlug="pressable"
+					isAlreadyOnSitesList={ true }
```
Comment (tyxla) [inline client/jetpack-connect/test/authorize.js]: Could be a good opportunity to clean those redundant `={ true }` prop values up:

```suggestion
					isAlreadyOnSitesList
```

File: client/jetpack-connect/test/authorize.js [hunk 6]
```diff
@@ -304,46 +354,107 @@ describe( 'JetpackAuthorize', () => {
...
+					isFetchingSites={ true }
```
Comment (tyxla) [inline client/jetpack-connect/test/authorize.js]:
```suggestion
					isFetchingSites
```

File: client/jetpack-connect/test/authorize.js [hunk 7]
```diff
@@ -304,46 +354,107 @@ describe( 'JetpackAuthorize', () => {
...
+				/>
+			);
+
+			await userEvent.click( screen.getByText( 'Return to your site' ) );
+
+			const target = global.window.location.href;
```
Comment (tyxla) [inline client/jetpack-connect/test/authorize.js:386]: Should `global.window.location` be a mock and then we could use `toHaveBeenCalledWith()` on it?

**Review by tyxla — COMMENTED** [review]
This looks great overall! Thanks for working on it ❤️ 

I've left a few suggestions for improvement, let me know what you think.

I think it's pretty close to go 🚀

File: client/jetpack-connect/test/authorize.js (hunk 5 above)
Comment (fullofcaffeine) [reply client/jetpack-connect/test/authorize.js]: 👍🏻

File: client/jetpack-connect/test/authorize.js (hunk 6 above)
Comment (fullofcaffeine) [reply client/jetpack-connect/test/authorize.js]: 👍🏻

File: client/jetpack-connect/test/authorize.js (hunk 4 above)
Comment (fullofcaffeine) [reply client/jetpack-connect/test/authorize.js:359]: > Also, why do we need to delete it if we overwrite it on the next line?

The `location` property is read-only, so we can't replace it with a new one without deleting it first: https://developer.mozilla.org/en-US/docs/Web/API/Window/location.

> I might be missing something, but since global.window.location is an object, originalWindowLocation will lead to it, and if we delete global.window.location, it will also erase originalWindowLocation, won't it?

It's indeed a bit counter-intuitive, but `delete` doesn't delete the referenced object (though it might be released by GC if it's not referenced anywhere else), but the _property_ itself, see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete:

> The delete operator removes a property from an object. If the property's value is an object and there are no more references to the object, the object held by that property is eventually released automatically.

I've recorded two short screencasts for clarity:


https://user-images.githubusercontent.com/81248/196293615-06a5c3f4-964e-4740-93e7-76f6700bc4b0.mp4


https://user-images.githubusercontent.com/81248/196293622-02f935e5-4419-4463-82ef-980f9f5972d1.mp4

File: client/jetpack-connect/test/authorize.js (hunk 7 above)
Comment (fullofcaffeine) [reply client/jetpack-connect/test/authorize.js:386]: Since we're checking the actual value of the `href` property (which is what is set directly by the given logic branch in the `navigate` function, see [#](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/navigate/index.ts#L14)), I don't understand how using `toHaveBeenCalledWith` here would work in this case 🤔. Could you provide an example?

**Comment by fullofcaffeine** [comment]
@tyxla Thank you for the review! I committed your commit suggestions and replied to your other comments.

File: client/jetpack-connect/test/authorize.js (hunk 3 above)
Comment (fullofcaffeine) [reply client/jetpack-connect/test/authorize.js]: 👍🏻

File: client/jetpack-connect/test/authorize.js (hunk 4 above)
Comment (tyxla) [reply client/jetpack-connect/test/authorize.js:359]: Interesting! Thanks for sharing your experience! I'd expect partially different behavior since it's not a native browser experience, but rather JSDOM. I'm pleasantly surprised that we get that level of parity between them 😉

File: client/jetpack-connect/test/authorize.js (hunk 7 above)
Comment (tyxla) [reply client/jetpack-connect/test/authorize.js:386]: Oh, I think I understand your confusion. I achieved this by using mocks with [getters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get) and [setters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/set) in a recent Gutenberg RTL migration. See #44470 and here's the actual mock [being declared](https://github.com/WordPress/gutenberg/blob/a3036828a41b0be8c79a6e7c459160d75cc018dd/packages/editor/src/components/post-preview-button/test/index.js#L30). At that point, it's easy enough to use `toHaveBeenCalledWith` on the inner mock that we invoke when the `location` or `href` setter is called. Let me know it that makes sense!

File: client/jetpack-connect/test/authorize.js [hunk 8]
```diff
@@ -352,13 +461,22 @@ describe( 'JetpackAuthorize', () => {
...
-			const target = component.instance().getRedirectionTarget();
+		test( 'should redirect to the /jetpack/connect/plans page by default', async () => {
+			renderWithRedux(
+				<JetpackAuthorize
+					{ ...DEFAULT_PROPS }
+					authQuery={ {
+						...DEFAULT_PROPS.authQuery,
+						alreadyAuthorized: true,
+					} }
+					isAlreadyOnSitesList={ true }
```
Comment (tyxla) [inline client/jetpack-connect/test/authorize.js]: Feel free to clean up some more of those redundant `={ true }` while we're here. There are more in the original code than my review pointed out. 😉

**Review by tyxla — APPROVED** [review]
I'd personally prefer the approach with mocks and setters and getters when mocking `window.location`, but that's not a requirement. Feel free to give it a try if you'd like.

Other than that, this looks great! Really nice work 🚀

File: client/jetpack-connect/test/authorize.js (hunk 7 above)
Comment (fullofcaffeine) [reply client/jetpack-connect/test/authorize.js:386]: That's neat! Thanks for the example! However, I think that for this specific test, the current approach ends up being more suitable/simpler because we still need to `delete` the `location` property to re-assign it, so `mockRestore` won't work on it, negating some (if not most) of the benefits of using the mock AFAICS. I'm happy to further discuss and revisit it in a follow-up, though!
